docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md

```diff
@@ -0,0 +1,378 @@
+# FileSystemStore in Mbed OS
+
+- [FileSystemStore in Mbed OS](#filesystemstore-in-mbed-os)
+    + [Revision history](#revision-history)
+- [Introduction](#introduction)
+    + [Overview and background](#overview-and-background)
+    + [Requirements and assumptions](#requirements-and-assumptions)
+- [System architecture and high-level design](#system-architecture-and-high-level-design)
+  * [Design basics](#design-basics)
+- [Detailed design](#detailed-design)
+    + [Class header](#class-header)
+    + [Important data structures](#important-data-structures)
+    + [Initialization and reset](#initialization-and-reset)
+    + [Core APIs](#core-apis)
+    + [Incremental set APIs](#incremental-set-apis)
+    + [Key iterator APIs](#key-iterator-apis)
+- [Usage scenarios and examples](#usage-scenarios-and-examples)
+    + [Standard usage of the class](#standard-usage-of-the-class)
+- [Other information](#other-information)
+    + [Open issues](#open-issues)
+
+
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
+| 1.0      	| 20 September 2018	| David Saada ([@davidsaada](https://github.com/davidsaada/)) 	| 5.11+           	| Initial revision 	|
+
+# Introduction
+
+### Overview and background
+
+FileSystemStore is a lightweight implementation of the [KVStore](../KVStore/KVStore_design.md) interface over file systems.
+
+### Requirements and assumptions
+
+FileSystemStore assumes the underlying file system qualities for resilience and file validation. This means that if the underlying file system has no protection against power failures, then neither would FileSystemStore have.  
+When initializing this class, it is assumed that the underlying FileSystem is initialized and mounted. 
+
+# System architecture and high-level design
+
+## Design basics
+
+FileSystemStore implements the get/set interface using files, where each key is represented by a single file. Key is represented by the file name and value is stored as file data. Therefore, FileSystemStore imitates the get/set actions using simple file operations. set is achieved using open-write-close, get using open-read-close and so on.  
+All files are concentrated under a single directory, whose name is hard coded. So actions such as "reset" are mapped to the deletion of all files under this directory, while iteration actions use file system APIs to traverse the directory.  
+
+### Data layout
+
+When storing the data, it is stored with a preceding 16-byte metadata header. Metadata includes flags and other parameters for basic validity checks.  
```

Comment:
Right. We first need to decide whether we want the write once feature or not for this class. It is now included for compatibility with `TDBStore` and `SecureStore`.

Reply:
@geky 
We need to understand "All of the supported filesystems actually support the WRITE_ONCE flag" because we do not know how to do that.
This flag is useful for having uniform tests over all implementations of kv_store interface. I do not see that somebody will use files in external devices that were encrypted in the device, therefore I do not see that adding this metadata has any bad side effect.

Reply:
That's true in the case of SecureStore, and I don't see a downside to adding metadata in SecureStore. But SecureStore is an optional layer. Outside of SecureStore, it's completely reasonable for a user to mount a file system as a KVStore for the simpler API or to achieve some sort of compatibility.

Reply:
I'm not sure I understand how write once can be supported in current file systems. Furthermore, I don't think we can count on future file systems to support it. Even if we decide not to support it in FileSystemStore, having no metadata means we can't support ANY flags in the future, if required to.

Reply:
One nice thing about flags is that we can always add features (until we run out of bits). If a user opens/sets a file with a flag the filesystem doesn't support, it should return an error.

We already have flags with partial support, for example the AUTHENTICATE_FLAG. If a user didn't read the documentation they may pass AUTHENTICATE_FLAG to TDBStore, in which case it should probably return an error.

If a feature isn't available in a specific filesystem, that's a limitation of the filesystem and the user should chose a different filesystem with the feature they're looking for.

Tangentially: We should add something like this to the handbook. Wikipedia has this for their big-ol-table of filesystems. Would probably help users:

| Storage Solution | KVStore API | FileSystem API | Power-loss resilience | Dynamic Wear-leveling | Static Wear-leveling | PC Support | Write-once | Encryption | Authentication | Rollback protection |
|------------------|-------------|----------------|-----------------------|-----------------------|----------------------|------------|------------|------------|----------------|---------------------|
| TDBStore | ✔️ | ❌ | ✔️ | ✔️ | ✔️ | ❌ | ✔️ | ❌ | ❌ | ❌ |
| LittleFS | ✔️ | ✔️ | ✔️ | ✔️ | ❌ | ❌ | ✔️ | ❌ | ❌ | ❌ |
| FAT | ✔️ | ✔️ | ❌ | ❌ | ❌ | ✔️ | ✔️ | ❌ | ❌ | ❌ |
| SecureStore | ✔️ | ❌ | ❌ | ❌ | ❌ | ❌ | ✔️ | ✔️ | ✔️ | ✔️ |

---

We don't have to add WRITE_ONLY right now, we could error on storage that isn't TDBStore or SecureStore, and add WRITE_ONLY in 5.12.

Though WRITE_ONLY support shouldn't be that difficult given that both FAT and LittleFS have on disk structures that support it (though it's not exposed in the FileSystem API yet).

FAT:
``` cpp
FATFileSystem::file_open(fs_file_t *file, const char *path, int flags) {
    // blablabla
    f_open(fh, fpath, openmode);
    if (flags & WRITE_ONCE) {
        f_chmod(fpath, AM_RDO, AM_RDO);
    }
    // blablabla
}
```

LittleFS:
``` cpp
LittleFileSystem::file_open(fs_file_t *file, const char *path, int flags) {
    // blablabla
    f->cfg = NULL;
    if (flags & WRITE_ONCE) {
        // this could be more efficient, though may be less readable
        uint32_t *mask = new uint32_t;
        *mask = 0444;
        f->cfg = new lfs_file_config;
        f->cfg->attrs = new lfs_attr;
        f->cfg->attrs->type = 0xff;
        f->cfg->attrs->buffer = mask;
        f->cfg->attrs->size = sizeof(uint32_t);
        f->cfg->attrs->next = NULL;
    }
    lfs_file_opencfg(&_lfs, f, path, f->cfg);
    // blablabla
}
```

Reply:
> If a user didn't read the documentation they may pass AUTHENTICATE_FLAG to TDBStore, in which case it should probably return an error.

I'm pretty sure I wrote somewhere in the doc (will look for it), but any unknown flag is ignored. This is important for KVStore stacking (all flags are transferred to the lower level as is).
Regarding the table - I like it, but this one is misleading. SecureStore over FAT is not power-loss resilient while over TDBStore is.